### ISSUE TITLE
[STORM-599] UI is slow due to fetching heartbeats from ZK

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#4440]
<p>The method getTopologyInfo fetches every heartbeat from ZooKeeper in order to gather statistics.  The UI calls this method via thrift to render the topology page.</p>

<p>When topologies are launched with thousands of executors, this fetching from ZK dramatically slows things down.</p>

---
<details><summary><i>Originally reported by <a title="Derek Dagit" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dagit">dagit</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-599" target="_blank">UI is slow due to fetching heartbeats from ZK</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Derek Dagit" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dagit">dagit</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2015-01-20T18:47:50+00:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/356#issuecomment-67997939" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/356#issuecomment-67997939</a></p>

<p>    The changes look fine to me.  My only concern would be around what happens if the cache has not been updated and is empty, but it looks like it will return an empty map, and that looks like what happened before when there were no heartbeats.</p>

<p>    +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/356" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/356</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>Thanks <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dagit" class="user-hover" rel="dagit">Derek Dagit</a>,</p>

<p>I merged this into master.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha">sriharsha</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dagit" class="user-hover" rel="dagit">Derek Dagit</a> This patch causing some issue in UI. Metrics don't show up in the UI after deploying wordcount topology.  I am able to view metrics after I reverted 40be5ddb94b674f8fcf2c804c7af96fbfe77776a, 3907b2d780b796e754bf3566d29567cff6e77679. <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=parth.brahmbhatt" class="user-hover" rel="parth.brahmbhatt">Parth Brahmbhatt</a> also verified that metrics are not showing up with latest master branch. Can you please check.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha" class="user-hover" rel="sriharsha">Harsha</a><br/>
You are correct I verified it too.  Hmm, we are running with this on based on a different code base that is working, it could have something to do with how often the cache is being updated.  I will revert the checking, reopen this and then we can investigate further.  Thanks for finding this.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>As per <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha" class="user-hover" rel="sriharsha">Harsha</a>'s comments this has been reverted, and we will explore what is happening.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user d2r opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/381" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/381</a></p>

<p>    <span class="error">&#91;Storm 599&#93;</span> Use use nimbus's cached heartbeats rather than fetching again from ZK, take two</p>

<p>    This is a retry of the reverted pull request #356.</p>

<p>    In #356, a merge error (we think?) on the branch used for the pull request caused part of the change on one line to go missing.</p>

<p>    ```Diff</p>
<ul class="alternate" type="square">
	<li>beats (.executor-beats storm-cluster-state storm-id (:executor->node+port assignment))<br/>
    + beats (map-val :heartbeat @(:heartbeats-cache nimbus))<br/>
    ```</li>
</ul>


<p>    The new code should have been:<br/>
    ```Diff<br/>
    + beats (map-val :heartbeat (get @(:heartbeats-cache nimbus) storm-id))<br/>
    ```</p>

<p>    This caused incorrect retrieval of heartbeats and therefore metrics were not populated as noted by @harshach <span class="error">&#91;here&#93;</span>(<a href="https://issues.apache.org/jira/browse/STORM-599?focusedCommentId=14275512&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14275512" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-599?focusedCommentId=14275512&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14275512</a>).</p>

<p>    Testing again with this branch in a clean environment shows that metrics are updated as expected once per `nimbus.monitor.freq.secs`, which defaults to 10s.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/d2r/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/d2r/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-599" title="UI is slow due to fetching heartbeats from ZK" class="issue-link" data-issue-key="STORM-599"><del>STORM-599</del></a>-use-cached-hbs</p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/381.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/381.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #381</p>

<hr />
<p>commit 1d973c6838118e64024505269b82baffd0f41f8d<br/>
Author: Derek Dagit <derekd@yahoo-inc.com><br/>
Date:   2015-01-14T17:12:30Z</p>

<p>    use hb cache instead of fetching via zk</p>

<p>commit fc952d4c64f52e29cd162ffacee984ef36ab9447<br/>
Author: Derek Dagit <derekd@yahoo-inc.com><br/>
Date:   2015-01-14T17:13:05Z</p>

<p>    Remove unnecessary code</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>The fixed version is not in master again.</p>
